src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -348,16 +417,350 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
+                {
+                    string odataId = resourceWrapper.ResourceBase.Id.OriginalString;
+
+                    IODataIdContainer container = new ODataIdContainer();
+                    container.ODataId = odataId;
+
+                    if (resource is EdmEntityObject edmObject)
+                    {
+                        edmObject.ODataIdContainer = container;
+                        edmObject.ODataPath = ODataPath;
+                    }
+                    else if (resource is IDeltaSetItem deltasetItem)
+                    {
+                        deltasetItem.ODataIdContainer = container;
+                        deltasetItem.ODataPath = ODataPath;
+                    }
+                    else
+                    {
+                        PropertyInfo containerPropertyInfo = EdmLibHelpers.GetClrType(entityType, readContext.Model).GetProperties().Where(x => x.PropertyType == typeof(IODataIdContainer)).FirstOrDefault();
+                        if (containerPropertyInfo != null)
+                        {
+                            IODataIdContainer resourceContainer = containerPropertyInfo.GetValue(resource) as IODataIdContainer;
+                            if (resourceContainer != null)
+                            {
+                                containerPropertyInfo.SetValue(resource, resourceContainer);
+                            }
+                            else
+                            {
+                                containerPropertyInfo.SetValue(resource, container);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private static Routing.ODataPath ApplyIdToPath(ODataDeserializerContext readContext, ODataResourceWrapper resourceWrapper)
+        {
+            // If an odata.id is provided, try to parse it as an OData Url.
+            // This could fail (as the id is not required to be a valid OData Url)
+            // in which case we fall back to building the path based on the current path and segments.
+            if (resourceWrapper.ResourceBase.Id != null)
+            {
+                try
+                {
+                    IWebApiRequestMessage internalRequest = readContext.InternalRequest;
+                    IWebApiUrlHelper urlHelper = readContext.InternalUrlHelper;
+
+                    string serviceRoot = urlHelper.CreateODataLink(
+                        internalRequest.Context.RouteName,
+                        internalRequest.PathHandler,
+                        new List<ODataPathSegment>());
+
+                    ODataUriParser parser = new ODataUriParser(readContext.Model, new Uri(serviceRoot), resourceWrapper.ResourceBase.Id);
+
+                    ODataPath odataPath = parser.ParsePath();
+                    if (odataPath != null)
+                    {
+                        return new Routing.ODataPath(odataPath);
```

Comment:
`parser.ParsePath` returns `ODataPath` from `Microsoft.OData.UriParser` namespace. This logic casts it to `Routng.ODataPath`